### PR TITLE
Add errors.Ignore()

### PR DESCRIPTION
**Pull request by camdencheek**
This adds the new helper function `Ignore` to our `errors` package. This
allows us to ignore errors based on a given predicate function in a way
that takes into account aggregated errors in MultiError. It recursively
unwraps to multierrors an filters out all child errors that match the
predicate.

## Test plan

Solid unit tests were added as part of this PR.

### CODE REVIEW

File: internal/search/run/expression_job.go [hunk 1]
```diff
@@ -161,14 +161,8 @@ func (j *OrJob) Run(ctx context.Context, db database.DB, stream streaming.Sender
 		})
 	}
 
-	// TODO(@camdencheek): errors.Is isn't good enough here since a single
-	// backend that returns a context.Canceled error will make the multierror
-	// return true for errors.Is(err, context.Canceled). Ideally, we have some
-	// sort of multi-error filter that can filter out any context.Canceled and
-	// leave us with whatever errors are left. Note that this is true of anywhere
-	// we check the type of an aggregated error. This is neither a new nor a
-	// unique problem.
```
Comment (camdencheek) [inline internal/search/run/expression_job.go:170]: This comment explains why this is needed.

**Review by efritz — APPROVED** [review]

**Review by eseliger — APPROVED** [review]

**Comment by unknwon** [comment]
Wow... when could we make this package to have the import path `github.com/sourcegraph/errors`?

**Comment by camdencheek** [comment]
>  I wanna use it outside the Sourcegraph.

The `./lib` directory has its own `go.mod` file, so you probably can use it outside of Sourcegraph if you want! I think the purpose of the `./lib` dir is exactly that: put code that we might want to use outside of the repo there.

**Comment by unknwon** [comment]
> > I wanna use it outside the Sourcegraph.
> 
> The `./lib` directory has its own `go.mod` file, so you probably can use it outside of Sourcegraph if you want! I think the purpose of the `./lib` dir is exactly that: put code that we might want to use outside of the repo there.

Thanks, that's a good one!

**Pull request merged**